asterixdb/asterix-algebra/src/main/java/org/apache/asterix/optimizer/rules/SpatialJoinRule.java

```diff
@@ -0,0 +1,247 @@
+                new MutableObject<>(new VariableReferenceExpression(leftInputVar)),
+                new MutableObject<>(new VariableReferenceExpression(rightInputVar)),
+                new MutableObject<>(new ConstantExpression(new AsterixConstantValue(
+                        new ARectangle(new APoint(spatialJoinAnn.getMinX(), spatialJoinAnn.getMinY()),
+                                new APoint(spatialJoinAnn.getMaxX(), spatialJoinAnn.getMaxY()))))),
+                new MutableObject<>(
+                        new ConstantExpression(new AsterixConstantValue(new AInt64(spatialJoinAnn.getNumRows())))),
+                new MutableObject<>(
+                        new ConstantExpression(new AsterixConstantValue(new AInt64(spatialJoinAnn.getNumColumns())))));
+
+        // Update the join conditions with the tile Id equality condition
+        ScalarFunctionCallExpression tileIdEquiJoinCondition =
+                new ScalarFunctionCallExpression(BuiltinFunctions.getBuiltinFunctionInfo(BuiltinFunctions.EQ),
+                        new MutableObject<>(new VariableReferenceExpression(leftTileIdVar)),
+                        new MutableObject<>(new VariableReferenceExpression(rightTileIdVar)));
+        ScalarFunctionCallExpression referenceIdEquiJoinCondition =
+                new ScalarFunctionCallExpression(BuiltinFunctions.getBuiltinFunctionInfo(BuiltinFunctions.EQ),
+                        new MutableObject<>(new VariableReferenceExpression(leftTileIdVar)),
+                        new MutableObject<>(referenceTileId));
+        //        ScalarFunctionCallExpression spatialJoinCondition = new ScalarFunctionCallExpression(
+        //                BuiltinFunctions.getBuiltinFunctionInfo(BuiltinFunctions.SPATIAL_INTERSECT),
+        //                new MutableObject<>(new VariableReferenceExpression(leftInputVar)),
+        //                new MutableObject<>(new VariableReferenceExpression(rightInputVar)));
+
+        conditionExprs.add(new MutableObject<>(spatialJoinFuncExpr));
+        conditionExprs.add(new MutableObject<>(tileIdEquiJoinCondition));
+        conditionExprs.add(new MutableObject<>(referenceIdEquiJoinCondition));
+
+        ScalarFunctionCallExpression updatedJoinCondition = new ScalarFunctionCallExpression(
+                BuiltinFunctions.getBuiltinFunctionInfo(BuiltinFunctions.AND), conditionExprs);
+        joinConditionRef.setValue(updatedJoinCondition);
+
+        List<LogicalVariable> keysLeftBranch = new ArrayList<>();
+        keysLeftBranch.add(leftTileIdVar);
+        keysLeftBranch.add(leftInputVar);
+        List<LogicalVariable> keysRightBranch = new ArrayList<>();
+        keysRightBranch.add(rightTileIdVar);
+        keysRightBranch.add(rightInputVar);
+        ISpatialJoinUtilFactory mjcf = new IntersectSpatialJoinUtilFactory();
+        joinOp.setPhysicalOperator(new SpatialJoinPOperator(joinOp.getJoinKind(),
```

Comment:
The `SpatialJoinPOperator` propably should take in the variables for the join and the variables for the tiles. This way you can make decisions on those variable in the join.

Reply:
I'm passing keysLeftBranch and keysRightBranch, which are list of variables (i.e., tileId and spatial field) to the constructor of SpatialJoinPOperator. Does this what you mean?